### ISSUE TITLE
A result type of `__builtin_choose_expr` that returns a pointer to a function is not correctly inferred in clang mode

### ISSUE THREAD

**Issue by mame** [microsoft/vscode-cpptools#9368]
Bug type: Language Service

**Describe the bug**
- OS and Version: Windows 11
- VS Code Version: 1.67.2
- C/C++ Extension Version: 1.9.8
- Other extensions you installed: Nothing
- If using SSH remote, specify OS of remote machine: Ubuntu on WSL
- When `__builtin_choose_expr` returns a pointer to a function, the result type seems not to be inferred correctly.

**Steps to reproduce**
1. Save the code sample as a file and open it in VSCode.
2. You will see an error `expression preceding parentheses of apparent call must have (pointer-to-) function type`

**Expected behavior**

No error is shown.

**Code sample and logs**
- Code sample

```
int foo(void) {
    puts("foo");
}

int bar(void) {
    puts("bar");
}

int main(void) {
    __builtin_choose_expr(1, foo, bar)();
}
```

**Screenshots**

<img width="609" alt="image" src="https://user-images.githubusercontent.com/21557/170426166-03af76ed-e15f-4980-b303-63ae159155e1.png">
NOTE: Screenshot is incorrect and missing an `_`.

**Additional context**

This idiom is heavily used in the source code of the programming language Ruby.

https://github.com/ruby/ruby/blob/2556e15d7520913119175c8166b1f9d17eb33415/include/ruby/internal/anyargs.h#L287

`RBIMPL_ANYARGS_DISPATCH_rb_define_method((arity), (func))` returns a pointer to a function by using `__builtin_choose_expr`, and then the function is invoked.

**Comment by sean-mcmanus** [comment]
The bug doesn't repro when the missing `_` is added -- are you still reproing this with `__builtin_choose_expr`? If so, can you give the exact compiler and cStandard setting that is being used for the repro and/or output of C/C++: Log Diagnostics?

**Comment by mame** [comment]
Thank you for your reply. Sorry, I seem to have lost the underscore when taking the screenshot. The issue still reproduces even with `__builtin_choose_expr` as follows.

<img width="505" alt="image" src="https://user-images.githubusercontent.com/21557/171313269-19ba3589-4e69-4986-a751-5f00141e51a9.png">

The example project has no `c_cpp_properties.json`. I created the file as follows, and the issue is still reproduced.

```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [],
            "compilerPath": "/usr/bin/clang",
            "cStandard": "c17",
            "cppStandard": "c++14",
            "intelliSenseMode": "linux-clang-x64"
        }
    ],
    "version": 4
}
```

Here is the output of C/C++: Log diagnostics

```
-------- Diagnostics - 6/1/2022, 11:15:36 AM
Version: 1.9.8
Current Configuration:
{
    "name": "Linux",
    "includePath": [
        "${workspaceFolder}/**"
    ],
    "defines": [],
    "compilerPath": "/usr/bin/clang",
    "cStandard": "c17",
    "cppStandard": "c++14",
    "intelliSenseMode": "linux-clang-x64",
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "compilerArgs": [],
    "mergeConfigurations": false,
    "browse": {
        "path": [
            "${workspaceFolder}/**"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}
Translation Unit Mappings:
[ /home/mame/work/ruby/t.c ]:
    /home/mame/work/ruby/t.c
Translation Unit Configurations:
[ /home/mame/work/ruby/t.c ]:
    Process ID: 417522
    Memory Usage: 11 MB
    Compiler Path: /usr/bin/clang
    Includes:
        /usr/lib/llvm-14/lib/clang/14.0.0/include
        /usr/local/include
        /usr/include/x86_64-linux-gnu
        /usr/include
    Standard Version: c17
    IntelliSense Mode: linux-clang-x64
    Other Flags:
        --clang
        --clang_version=140000
Total Memory Usage: 11 MB

------- Workspace parsing diagnostics -------
Number of files discovered (not excluded): 27641
Number of files parsed: 1

```

FYI: This is a screenshot of the issue with a working real-world application (Ruby)

<img width="537" alt="image" src="https://user-images.githubusercontent.com/21557/171314868-ac650a5a-6ae1-4642-9322-ec2c4d7f92b3.png">

https://github.com/ruby/ruby/blob/936e54063d4de0e44925e9139cb4d25cce27c187/array.c#L8746-L8779

`rb_define_method` is a macro, and the expanded code uses `__builtin_choose_expr` that returns a function pointer.

**Comment by sean-mcmanus** [comment]
This is the "same" issue as https://github.com/microsoft/vscode-cpptools/issues/8294, which got fixed for gcc mode but not clang mode.

**Comment by mame** [comment]
Thank you! I confirmed the errors dissapper if I change the `compilerPath` to `"/usr/bin/gcc"` in c_cpp_properties.json.

**Comment by sean-mcmanus** [comment]
This is fixed in 1.13.4 (pre-release): https://github.com/microsoft/vscode-cpptools/releases/tag/v1.13.4, although there's a related issue at https://github.com/microsoft/vscode-cpptools/issues/10168 that won't be fixed till 1.14.0.
